Add plugin library for ratsd

cowbon opened the pull request:
Add the plugin framework similar to Veraison/service for sub-attesters. Unlike services, ratsd plugins communicates with ratsd core through GRPC,and there is only one type of the plugin. The verification for plugins will be added in the later patches.

jraman567 commented on cmd/main.go:
```diff
@@ -60,6 +62,23 @@ func main() {
+
+	pluginManager, err := plugin.CreateGoPluginManagerWithLoader(
+		pluginLoader, log.Named("builtin"))
+
```
Can you replace the calls to `plugin.CreateGoPluginLoader()` & `plugin.CreateGoPluginManagerWithLoader()` with a single call to `plugin.CreateGoPluginManager()` ?

cowbon replied:
Agree, there's currently no `builtin` loader here in ratsd as opposed to `veraison/services`. We might want to add a `secure loader` in ratsd as the option in the future, and we'll have to separate the loader and the manager again then.

jraman567 reviewed (commented):
Thanks for the patches, @cowbon. I left a few nitpicking comments; please address them.

The PR looks good to me by and large.

I got confused by the use of "plugin" in the code. In some cases (in main.go), that refers to the plugin you implemented, but it relates to the go-plugin library within the plugin directory - just a note.

I got a chance to compare it against Veraison's implementation, and they look similar. LGTM!

cowbon commented on plugin/goplugin_context.go:
```diff
@@ -0,0 +1,93 @@
+type PluginContext struct {
+	// Path to the exectable binary containing the plugin implementation
+	Path string
+	// Name of this plugin
```
Added the line after line 19.

cowbon commented on plugin/goplugin_context.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+// PluginConntext is a generic for handling ratsd plugins. It is
+// parameterised on the IPluggale interface it handles.
```
Fixed.

cowbon commented on plugin/goplugin_context.go:
```diff
@@ -0,0 +1,93 @@
+	"go.uber.org/zap"
+)
+
+// PluginConntext is a generic for handling ratsd plugins. It is
```
Removed additional alphabet "n" from `PluginConntext`. It should have the right spelling now.

cowbon commented:
Hi @jraman567 Thanks for the review, note the implementations of the ratsd plugin will be called `attesters`. There will be another directory named as such to hold the attesters, e.g. TSM.

jraman567 reviewed (approved):
@cowbon The PR looks good to me.

thomas-fossati commented on plugin/imanager.go:
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: Apache-2.0
+package plugin
+
+type IManager interface {
```
This should have a godoc comment

cowbon replied:
Added the GoDoc similar to `veraison/service/plugin`

thomas-fossati commented on plugin/ipluggable.go:
```diff
@@ -0,0 +1,16 @@
+// IPluggable respresents a "pluggable" point within Veraison ratsd.
+type IPluggable interface {
+	GetEvidence(in *compositor.EvidenceIn) *compositor.EvidenceOut
+
```
Spurious newline

thomas-fossati commented on plugin/ipluggable.go:
```diff
@@ -0,0 +1,16 @@
+	GetEvidence(in *compositor.EvidenceIn) *compositor.EvidenceOut
+
+	GetSubAttesterID() *compositor.SubAttesterIDOut
+
```
ditto

cowbon replied:
Same as above

thomas-fossati commented on plugin/goplugin_rpc.go:
```diff
@@ -0,0 +1,80 @@
+	resp, err := c.client.GetSubAttesterID(context.Background(), &emptypb.Empty{})
+	if err != nil {
+		return &compositor.SubAttesterIDOut{
+			Status: &compositor.Status{Result: true, Error: err.Error()},
```
On error, status.result should be false

cowbon replied:
Switched the logic to `Result: false` in the latest revision.

thomas-fossati reviewed (commented):
LGTM, thanks!

I have left a couple of simple-to-address comments inline.

cowbon commented:
Thanks @thomas-fossati for your comments. I've added the GoDoc in the similar fashion of the `plugin` in `services`, and changed the logic of `Status.Result` when an error occurs.

thomas-fossati reviewed (approved):
Thanks @cowbon !
